pkg/common/peertracker/npipe_windows.go

```diff
@@ -0,0 +1,63 @@
+//go:build windows
+// +build windows
+
+package peertracker
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"syscall"
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+)
+
+var (
+	kernelbase = windows.NewLazyDLL("kernelbase.dll")
+	kernel32   = windows.NewLazyDLL("kernel32.dll")
+
+	procCompareObjectHandles        = kernelbase.NewProc("CompareObjectHandles")
+	procGetNamedPipeClientProcessID = kernel32.NewProc("GetNamedPipeClientProcessId")
+)
+
+func getCallerInfoFromNamedPipeConn(conn net.Conn) (CallerInfo, error) {
+	var info CallerInfo
+
+	type Fder interface {
+		Fd() uintptr
+	}
+	fder, ok := conn.(Fder)
+	if !ok {
+		conn.Close()
+		return info, errors.New("invalid connection")
+	}
+
+	var pid int32
+	if err := getNamedPipeClientProcessID(windows.Handle(fder.Fd()), &pid); err != nil {
+		return info, fmt.Errorf("error in GetNamedPipeClientProcessId function: %w", err)
+	}
+
+	return CallerInfo{
+		Addr: conn.RemoteAddr(),
+		PID:  pid,
+	}, nil
+}
+
+func getNamedPipeClientProcessID(pipe windows.Handle, clientProcessID *int32) (err error) {
+	r1, _, e1 := syscall.Syscall(procGetNamedPipeClientProcessID.Addr(), 2, uintptr(pipe), uintptr(unsafe.Pointer(clientProcessID)), 0)
```

Comment:
Just a heads up that this will need to change to the new SyscallN function when we move to go1.18.